# Feature Extraction From Text

### Text Representation / Text Vectorization

- One Hot Encoding
- Bag of Words (BOW)
- Bag of n-grams
- Tf/Idf (Term Frequency / Inverse Document Frequency)
- Custom / Hybrid Features

In [2]:
import numpy as np
from collections import Counter
import math

from sklearn.feature_extraction.text import TfidfVectorizer

In [3]:
corpus = [
    "machine learning is fun",
    "machine learning and deep learning",
    "deep learning drives artificial intelligence",
    "artificial intelligence and machine learning",
    "natural language processing is part of artificial intelligence",
    "python is popular for machine learning",
    "data science includes machine learning",
    "statistics is important for data science",
    "deep learning uses neural networks",
    "neural networks are used in artificial intelligence"
]

# Custom TF-IDF Vectorizer

1. Build the vocab
2. Build a BOW (Bag of Words Vector)
3. Build a TF matrix (Term Frequency. How often the word appears in all the documents)
4. Build an IDF Matrix (How rare the word is througout the corpus)

    IDF of a Word = Total No of Docs / Total no of documents with that word  
  
    IDF(the) = 4 / 4 = 1 (Less Important as it appears in every document)  
    IDF(sky) = 4 / 2 = 2 (More important word than the)  

   - To avoid dividing by zero we add 1 in both numerator and denominator  
   - To normalize we take a log (Taking log will normalize. Let suppose we have a very large no of docs or a word appears hundreds of time)

5. Finally we compute TF-IDF matrix

## Build the Vocabulary

In [4]:
vocab = set()
for doc in corpus:
    for word in doc.split(" "):
        vocab.add(word.strip().lower())

print(f"Corpus Size: {len(vocab)}")
print(f"Vocabulary: {vocab}")

Corpus Size: 28
Vocabulary: {'networks', 'deep', 'popular', 'data', 'part', 'important', 'learning', 'in', 'is', 'python', 'are', 'and', 'science', 'intelligence', 'natural', 'used', 'includes', 'artificial', 'machine', 'language', 'statistics', 'processing', 'of', 'uses', 'neural', 'for', 'fun', 'drives'}


## Compute TF Vector per document for all the words in Vocab

In [5]:
# Initialize an array of size vocabular length
# Each index represent one word in the vocab
# Loop through all the docs in the corpus
# Then Loop through all the words in a doc
# for each word in vocab, the function adds the
# count of time that word appears in the vocab
def get_tf_matrix(corpus, vocab):
    word_to_index = {word: i for i, word in enumerate(vocab)}
    matrix = np.zeros((len(corpus), len(vocab)))
    
    for doc_idx, doc in enumerate(corpus):
        for word in doc.split():
            if word in word_to_index:
                matrix[doc_idx, word_to_index[word]] += 1
                
    return matrix

In [6]:
print(vocab)
tf_vector = get_tf_matrix(corpus, vocab)
tf_vector

{'networks', 'deep', 'popular', 'data', 'part', 'important', 'learning', 'in', 'is', 'python', 'are', 'and', 'science', 'intelligence', 'natural', 'used', 'includes', 'artificial', 'machine', 'language', 'statistics', 'processing', 'of', 'uses', 'neural', 'for', 'fun', 'drives'}


array([[0., 0., 0., 0., 0., 0., 1., 0., 1., 0., 0., 0., 0., 0., 0., 0.,
        0., 0., 1., 0., 0., 0., 0., 0., 0., 0., 1., 0.],
       [0., 1., 0., 0., 0., 0., 2., 0., 0., 0., 0., 1., 0., 0., 0., 0.,
        0., 0., 1., 0., 0., 0., 0., 0., 0., 0., 0., 0.],
       [0., 1., 0., 0., 0., 0., 1., 0., 0., 0., 0., 0., 0., 1., 0., 0.,
        0., 1., 0., 0., 0., 0., 0., 0., 0., 0., 0., 1.],
       [0., 0., 0., 0., 0., 0., 1., 0., 0., 0., 0., 1., 0., 1., 0., 0.,
        0., 1., 1., 0., 0., 0., 0., 0., 0., 0., 0., 0.],
       [0., 0., 0., 0., 1., 0., 0., 0., 1., 0., 0., 0., 0., 1., 1., 0.,
        0., 1., 0., 1., 0., 1., 1., 0., 0., 0., 0., 0.],
       [0., 0., 1., 0., 0., 0., 1., 0., 1., 1., 0., 0., 0., 0., 0., 0.,
        0., 0., 1., 0., 0., 0., 0., 0., 0., 1., 0., 0.],
       [0., 0., 0., 1., 0., 0., 1., 0., 0., 0., 0., 0., 1., 0., 0., 0.,
        1., 0., 1., 0., 0., 0., 0., 0., 0., 0., 0., 0.],
       [0., 0., 0., 1., 0., 1., 0., 0., 1., 0., 0., 0., 1., 0., 0., 0.,
        0., 0., 0., 0., 1

## Compute IDF for all the words

In [7]:
def compute_idf(corpus, vocab):
    N = len(corpus)
    df = Counter()

    for doc in corpus:
        for word in set(doc.split()):
            df[word] += 1

    idf = {
        word: math.log((N + 1) / (df[word] + 1))
        for word in df
    }

    return idf

In [8]:
idf = compute_idf(corpus, vocab)

idf

{'machine': 0.6061358035703155,
 'is': 0.7884573603642703,
 'learning': 0.3184537311185346,
 'fun': 1.7047480922384253,
 'and': 1.2992829841302609,
 'deep': 1.0116009116784799,
 'intelligence': 0.7884573603642703,
 'artificial': 0.7884573603642703,
 'drives': 1.7047480922384253,
 'natural': 1.7047480922384253,
 'language': 1.7047480922384253,
 'of': 1.7047480922384253,
 'processing': 1.7047480922384253,
 'part': 1.7047480922384253,
 'popular': 1.7047480922384253,
 'for': 1.2992829841302609,
 'python': 1.7047480922384253,
 'science': 1.2992829841302609,
 'includes': 1.7047480922384253,
 'data': 1.2992829841302609,
 'important': 1.7047480922384253,
 'statistics': 1.7047480922384253,
 'uses': 1.7047480922384253,
 'networks': 1.2992829841302609,
 'neural': 1.2992829841302609,
 'in': 1.7047480922384253,
 'used': 1.7047480922384253,
 'are': 1.7047480922384253}

# Using TfidfVectorizer

- Calculate TF-IDF score using Sklearn TfIdfVectorizer

In [9]:
vec = TfidfVectorizer()

## First let remove stop words and any punctuations

In [21]:
import spacy
nlp = spacy.load("en_core_web_sm")

In [22]:
def preprocess(text):
    doc = nlp(text)
    tokens = [token.lemma_ for token in doc if not token.is_stop and not token.is_punct]
    return (' '.join(tokens)).strip()

In [26]:
processed_corpus = [preprocess(text) for text in corpus]

processed_corpus

['machine learning fun',
 'machine learning deep learning',
 'deep learning drive artificial intelligence',
 'artificial intelligence machine learning',
 'natural language processing artificial intelligence',
 'python popular machine learning',
 'data science include machine learning',
 'statistic important datum science',
 'deep learning use neural network',
 'neural network artificial intelligence']

In [27]:
vec_output = vec.fit_transform(processed_corpus)

In [28]:
print(vec.vocabulary_)

{'machine': 11, 'learning': 10, 'fun': 5, 'deep': 3, 'drive': 4, 'artificial': 0, 'intelligence': 8, 'natural': 12, 'language': 9, 'processing': 16, 'python': 17, 'popular': 15, 'data': 1, 'science': 18, 'include': 7, 'statistic': 19, 'important': 6, 'datum': 2, 'use': 20, 'neural': 14, 'network': 13}


In [29]:
all_features = vec.get_feature_names_out()

for feature in all_features:
    idx = vec.vocabulary_.get(feature)
    print(f"{feature}:{vec.idf_[idx]}")

artificial:1.7884573603642702
data:2.7047480922384253
datum:2.7047480922384253
deep:2.01160091167848
drive:2.7047480922384253
fun:2.7047480922384253
important:2.7047480922384253
include:2.7047480922384253
intelligence:1.7884573603642702
language:2.7047480922384253
learning:1.3184537311185345
machine:1.6061358035703155
natural:2.7047480922384253
network:2.2992829841302607
neural:2.2992829841302607
popular:2.7047480922384253
processing:2.7047480922384253
python:2.7047480922384253
science:2.2992829841302607
statistic:2.7047480922384253
use:2.7047480922384253


- It's important to pre-process and remove stop words and reduce words to their base form.
- Than after that taking TfIdf you can see.
- The score of Python is higher (~2.70) as it appears lesser no of times.
- But the score for machine is lower (~1.60) as it appears more in the corpus.